File: leetcode/0140.word-break-2/go/001.backtracking.go
func wordBreak(s string, wordDict []string) []string {
    var res []string
    var substrs []string

    wordSet := make(map[string]bool)
    for _, word := range(wordDict) {
        wordSet[word] = true
    }

    var backtrack func(int)
    backtrack = func(i int) {
        if i == len(s) {
            res = append(res, strings.Join(substrs, " "))
            return
        }

        for j := i; j < len(s); j++ {
            word := s[i:j+1]
            if wordSet[word] {
                substrs = append(substrs, word)
                backtrack(j+1)
                substrs = substrs[:len(substrs)-1]
            }
        }
    }

    backtrack(0)
    
    return res
}

File: leetcode/0140.word-break-2/ts/001.backtracking.ts
function wordBreak(s: string, wordDict: string[]): string[] {
    const res = [];
    const substrs = [];
    const wordSet = new Set(wordDict);

    const backtrack = (i) => {
        if(i === s.length) {
            res.push(substrs.join(' '));
            return
        }

        for(let j = i; j < s.length; j++) {
            const word = s.substring(i, j+1);
            if(wordSet.has(word)) {
                substrs.push(word);
                backtrack(j+1);
                substrs.pop();
            }
        }
    }

    backtrack(0);
    return res;
};
